File: 프로그래머스/2/388353. 지게차와 크레인/지게차와 크레인.py
from collections import deque
def solution(store, rqs):
    arr = []
    ans = 0
    dx = [0,0,1,-1]
    dy = [1,-1,0,0]
    for s in store:
        row = list()
        for c in s:
            row.append(c)
        arr.append(row)
    n = len(arr)
    m = len(arr[0])
    def check(x,y):
        return 0 <= x < n and 0 <= y < m
    
    def bfs(x,y,map):
        q = deque()
        visit = [[False] * m for _ in range(n)]
        visit[x][y] = True
        q.append((x,y))
        while q:
            cx,cy = q.popleft()
            for i in range(4):
                nx,ny = cx + dx[i], cy + dy[i]
                if not check(nx,ny):
                    return True
                if not visit[nx][ny] and map[nx][ny] == 0:
                    visit[nx][ny] = True
                    q.append((nx,ny))
        return False
    
    def removeAll(con):
        for i in range(n):
            for j in range(m):
                if arr[i][j] == con:
                    arr[i][j] = 0
                
    def removeOutSide(con,arr):
        # 처음부터 바깥 혹은 0과 맞대고 있어야 함
        # 순서대로 치우다가 0을 맞댄 녀석이면 해당되지 않는다!
        # 이전과 비교를 할 map 복사
        map = [row[:] for row in arr]
        for i in range(n):
            for j in range(m):
                if arr[i][j] == con:
                    result = bfs(i,j,map)
                    if result:
                        arr[i][j] = 0
    
    for request in rqs:
        # request 길이가 1이면 4면이 바깥 or 0인 거
        if len(request) > 1:
            removeAll(request[0])
        else:
            removeOutSide(request[0],arr)
    for i in range(n):
        for j in range(m):
            if arr[i][j] != 0:
                ans += 1
    print(ans)
    return ans

File: 프로그래머스/2/388353. 지게차와 크레인/지게차와 크레인.java
import java.util.*;

class Solution {
    static final int[][] dir = {{0,1},{0,-1},{1,0},{-1,0}};
    static final char EMPTY = '.';
    static int n, m;

    public int solution(String[] storage, String[] req) {
        int ans = 0;
        n = storage.length;
        m = storage[0].length();

        char[][] arr = new char[n][m];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                arr[i][j] = storage[i].charAt(j);
            }
        }

        // 요청 처리
        for (String r : req) {
            if (r.length() == 1) {
                // 외부와 접촉 가능한 경로(EMPTY 또는 target)만 따라가며 target 제거
                char target = r.charAt(0);
                // 비교할 map(깊은 복사)
                char[][] map = new char[n][m];
                for (int i = 0; i < n; i++) map[i] = arr[i].clone();
                // 지점 판단
                for (int i = 0; i < n; i++) {
                    for (int j = 0; j < m; j++) {
                        if (arr[i][j] == target) {
                            if (bfs(i,j,target,arr,map)) arr[i][j] = EMPTY;
                        }
                    }
                }

            } else {
                // 해당 문자를 전부 제거
                char target = r.charAt(0);
                for (int i = 0; i < n; i++) {
                    for (int j = 0; j < m; j++) {
                        if (arr[i][j] == target) {
                            arr[i][j] = EMPTY;
                        }
                    }
                }
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (arr[i][j] != EMPTY) ans++;
            }
        }
        return ans;
    }

    static boolean bfs(int x, int y, char target, char[][] arr, char[][] map) {
        Deque<int[]> q = new ArrayDeque<>();
        boolean[][] visit = new boolean[n][m];
        q.add(new int[]{x,y});
        visit[x][y] = true;
        while(!q.isEmpty()) {
            int[] cur = q.poll();
            int cx = cur[0];
            int cy = cur[1];
            for (int[] d : dir) {
                int nx = cx + d[0];
                int ny = cy + d[1];
                if (!check(nx,ny)) return true;
                if (!visit[nx][ny] && map[nx][ny] == EMPTY) {
                    visit[nx][ny] = true;
                    q.add(new int[] {nx,ny});
                }
            }
        }
        return false;
    }
    
    public static boolean check(int x,int y) {
        return 0 <= x && x < n && 0 <= y && y < m;
    }
}